packages/flutter/lib/src/widgets/_web_image_web.dart

```diff
@@ -0,0 +1,344 @@
+        // The image loaded, so we need to clean up the cache now that we know
+        // its size.
+        _cleanupCache();
+        info!.state.removeListener(imgLoadListener!);
+      };
+      info.state.addListener(imgLoadListener);
+      _imageCache[url] = info;
+    }
+    _touch(url);
+    return info;
+  }
+
+  static void _touch(String url) {
+    _imageQueue.remove(url);
+    _imageQueue.addFirst(url);
+  }
+
+  /// Remove the least recently used items from the cache until the number of
+  /// items in the cache is less than [_maxImages] and the total bytes of the
+  /// images in the cache is less than [_maxSizeInBytes].
+  static void _cleanupCache() {
+    assert(_imageQueue.length == _imageCache.length);
+    if (_imageQueue.length > _maxImages) {
+      final int imagesToRemove = _imageQueue.length - _maxImages;
+      for (int i = 0; i < imagesToRemove; i++) {
+        final String url = _imageQueue.removeLast();
+        final _ImgElementProviderImpl? info = _imageCache.remove(url);
+        if (info != null) {
+          if (info.state.value == _ImageLoadingState.loading) {
+            info.state.dispose();
+          }
+        }
+      }
+    }
+
+    double totalSize = 0;
+    for (final String url in _imageQueue) {
+      final _ImgElementProviderImpl? info = _imageCache[url];
+      totalSize += info!.estimatedSize;
+    }
+    while (totalSize > _maxSizeInBytes) {
+      final String url = _imageQueue.removeLast();
+      final _ImgElementProviderImpl? info = _imageCache.remove(url);
+      if (info != null) {
+        if (info.state.value == _ImageLoadingState.loading) {
+          info.state.dispose();
+        }
+        totalSize -= info.estimatedSize;
+      }
+    }
+  }
+}
+
+web.HTMLImageElement _createImage(String url) {
+  final web.HTMLImageElement image =
+      web.document.createElement('img') as web.HTMLImageElement;
+  image.src = url;
+  image.style
+    ..width = '100%'
+    ..height = '100%';
+  return image;
+}
+
+enum _ImageLoadingState {
+  loading,
+  success,
+  error,
+}
+
+class _ImgElementProviderImpl extends ImgElementProvider {
+  _ImgElementProviderImpl(this.src) : image = _createImage(src) {
+    _startDecoding();
+  }
+
+  final String src;
+  final web.HTMLImageElement image;
+
+  ValueNotifier<_ImageLoadingState> state =
+      ValueNotifier<_ImageLoadingState>(_ImageLoadingState.loading);
+
+  late Object? _error;
+  Object get error {
+    assert(state.value == _ImageLoadingState.error);
+    return _error!;
+  }
+
+  late StackTrace? _stackTrace;
+  StackTrace get stackTrace {
+    assert(state.value == _ImageLoadingState.error);
+    return _stackTrace!;
+  }
+
+  Future<void> _startDecoding() async {
+    try {
+      await image.decode().toDart;
+      state.value = _ImageLoadingState.success;
+    } catch (e, stackTrace) {
+      state.value = _ImageLoadingState.error;
+      _error = e;
+      _stackTrace = stackTrace;
+    }
+    state.dispose();
```

Comment:
@harryterkelsen state.dispose() should not be here for sure, it contradicts with: await image.decode().toDart;

Reply:
Out of date. I refactored this.